src/amber.cc

```diff
@@ -86,6 +88,57 @@ amber::Result Amber::ExecuteWithShaderData(const amber::Recipe* recipe,
     return r;
   }
 
+  if (!opts.image_file.empty()) {
+
+    // Write framebuffer image to PPM file
+
+    ResourceInfo framebuffer_info;
+    r = engine->GetFrameBufferInfo(&framebuffer_info);
+    if (!r.IsSuccess()) {
+      return r;
+    }
+
+    if (framebuffer_info.type != ResourceInfoType::kImage) {
+      return Result("Resource is not an image");
+    }
+
+    const uint32_t width = framebuffer_info.image_info.width;
+    const uint32_t height = framebuffer_info.image_info.height;
+
+    FormatBuffer pixels(BufferType::kFramebuffer);
+    pixels.SetName("Framebuffer pixels");
+    r = engine->GetFrameBuffer(&pixels);
+    if (!r.IsSuccess()) {
+      return r;
+    }
+
+    if (pixels.GetFormat().GetFormatType() != FormatType::kR8G8B8A8_UINT) {
+      return Result("Only R8G8B8A8 is supported for now");
+    }
+
+    std::ofstream image_stream;
+    image_stream.open(opts.image_file, std::ios::out | std::ios::binary);
+    if (!image_stream.is_open()) {
+      return Result("Cannot open file to write image");
+    }
+
+    // PPM header
+    const uint32_t maximum_color_value = 255;
+    image_stream << "P6" << std::endl;
+    image_stream << width << " " << height << std::endl;
+    image_stream << maximum_color_value << std::endl;
+
+    // PPM data
+    for (Value data: pixels.GetData()) {
+      const uint32_t pixel = data.AsUint32();
+      const char *p = reinterpret_cast<const char *>(&pixel);
+      // Write only RGB channels, PPM does not support alpha
+      image_stream.write(p, 3);
+    }
+
+    image_stream.close();
+  }
+
   return engine->Shutdown();
```

Comment:
If we move the engine->Shutdown() to the destructor, then we can add a new call into amber to get the framebuffer instead of having to do it here, yea? We're going to want the ability to write out data buffers as well as framebuffers in the future so would be preferrable to have that code live elsewhere.

Reply:
> If we move the engine->Shutdown() to the destructor, then we can add a new call into amber to get the framebuffer instead of having to do it here, yea?

Just to clarify: you mean the destructor of the Amber object? Then engine must become a field of Amber, and be started/checked to be the right Engine whenever a flavor of Amber::Execute() is called?

Making the Engine survive an Execute call means that all Execute calls have to check that the maybe already started engine is of the type passed in the opts.engine argument of Execute, am I right?

Reply:
Yup, I agree it's crappy, heh. I like your idea of using options, though I think we need to extend it a bit as I may want to extract multiple buffers (data buffers or data and frame, etc).

Maybe something like:
```
struct BufferInfo {
   std::string buffer_name;
   std::vector<Value> values;
};
struct Options {
  EngineType engine;
  EngineConfig* config;

  vector<BufferInfo> extractions;
};
```

Then on the client side, you'd push on the bufferInfo objects you want to extract and execute would iterate over the list and write the contents of that buffer into the values list?

Does that seem like it would work?

Reply:
Yes that looks good! I'll give a go at this approach as soon as feasible. Thanks!